### PR TITLE
Overhaul of the test subsystem.

### LINKED ISSUE #331: Nose testing, plugins, and the use of numpy.testing
Sorry for the long post. Please bear with me because I need your guidance!

In trying to enhance testing to perform memleak detection (#326) I found that the only way to fully extend `nose` is through plugins.
Other approaches using `setUp` and `tearDown` fixtures might work but seem at the moment very hackish and too intrusive (from the discussion in #326: we want to be able to memleak test, possibly on a test-by-test basis, but we shouldn't be forced to do so every time we test nor for all tests.)

Now, `nose` plugins are great but in order for `nose` to use them they must be either:
1- A package that's installed system-wide (or at least in the user's package repository);
2- Loaded at runtime by the script that calls `nose.main()`.

I think we can all agree that option 1 is overkill and unnecessarily messy. Option 2 holds promise, but there is, of course, a catch: tests must be run from a script. This means that running `nosetests MDAnalysisTests` from the command-line won't load the plugin since no launch script will be run.

So, any advance in this front would require breaking the command-line usability of `nosetests` and testing would have to be done by running `MDAnalysis.tests.test()` from within a python script in which the plugin(s) are loaded. (`nosetests` is also just a python script, but a very simple one that can't load any extra plugins).

A similar issue happened with the `KnownFailure` plugin from `numpy`. We have the decorators in place, the plugin is there to output `K` instead of `S`, and the wiki even gives examples where the output has `K` and `KNOWNFAIL`. But commit 073b075dd3ec980eacd7920c3ea747cfe5ed18ca shows that we ditched that because `nosetests` bypasses the loading of `numpy`'s plugins.

What can we do? I see some ways out:
2.1- We break the use of `nosetests` from the command-line and freely use plugins;
2.2- We try to implement plugins in a way that, when `nosetests` is called from the command-line and the plugins aren't loaded tests still work fine, but with limited functionality (for instance, the `KnownFailure` plugin falling back to the `SKIP` behavior we have now).

Option 2.1 is simple to implement and doesn't even mean we can no longer run tests from the command-line. It's enough to write up a slightly smarter version of `nosetests` that preloads all our plugins. We could even call it `nosetests` just the same and have it live in our `MDAnalysisTests` package dir.
This also means that travis building shouldn't be affected; we need only run this script instead of the dumb `nosetests`.

Option 2.2 is probably the most friendly to all, but entails more work for current and future plugins. My idea is that we can set some flags in the testing environment depending on whether plugins are loaded, and then have the tests behave accordingly.

The options are more complementary than alternative: we can have option 2.2 and still have the script that fires up the plugin-enabled testing.

Personally, I lean more towards 2.1 but ultimately I think this depends on how our devs typically run the tests. So please, opinions!

### LINKED ISSUE #338: knownfailure decorator isn't always honored.
In trying to wrap up #331 I found out that the `@knownfailure` decorator sometimes causes a test to not be run (not to confuse with run but skipped if an exception occurs, which is the function of the decorator).

The following, in particular, doesn't work as expected:
Create a MDAnalysisTests/test_kf.py with the following
```python
from MDAnalysisTests import knownfailure
from numpy.testing import assert_, TestCase

class TestKF(TestCase):
    def test_1():
        pass

    @knownfailure
    def test_2():
        assert_(False)
```

running `nosetests MDAnalysisTests/test_kf.py` yields the following:
```
.
----------------------------------------------------------------------
Ran 1 test in 0.001s

OK
```
Notice that only one test was run and there's no mention to `test_2` being skipped. If I make nose very verbose I can see that `test_2` gets its name clobbered by the name of the function returned by the decorator (`knownfailure_decorator`), and therefore doesn't get selected:
```
nose.selector: DEBUG: wantMethod <unbound method TestKF.test_1>? True
nose.selector: DEBUG: wantMethod <unbound method TestKF.knownfailure_decorator>? None
```

I'm trying to track this down, and understand why it works on some occasions but not in others (after all, the decorator does work in other places in the test suite). It's not a very serious problem because the consequence is that we don't run a test that were going to skip anyway. We miss output, and tracking of tests that should fail but don't.

For reference I tested this in `develop` and `master`, with same results.

### PR DESCRIPTION
**Pull request by mnmelo**
Implements nose plugins, as a submodule (closes #331).

Available plugins are: memleak testing, stderr capturing
 (for quieter test runs), and proper knownfailure implementation.

The test function is no longer imported from numpy. We
 roll our own that registers our external plugins.

Behavior was kept regarding parallel testing, labels, and invocation
 from the command line. stdout and stderr suppression was made default.

Running nosetests from the command line bypasses our test function,
 and plugins aren't loaded. Plugin-dependent code (the knownfailure
 decorator, for instance) is aware of this and behaves accordingly.

A command-line script mda_nosetests was added to the suite that acts
 as a replacement for the nosetests command, but is plugin-friendly.

Knownfailure usage was corrected (closes #338).

(Replaces pull request #326)

### CODE REVIEW

**Comment by mnmelo** [comment]
Just some more explanation:
From the command-line, if using the `mda_nosetests` script, you can use `-s` to show stdout, and `--no-errorcapture`to show stderr.
`--with-memleak` activates test-by test memleak checking.
`--no-knownfail` deactivates special handling of `KnownFail` exceptions, which will then be counted as failures.

From direct invocation of `MDAnalysis.tests.test` all the above flags can be set if passed to `extra_argv`.

**Comment by orbeckst** [comment]
@mnmelo – excellent groundwork in the deep bowels of the unit tests! Once this is merged, can you add something on the use of the `mda_nosetests` script to the wiki page on [UnitTests](/MDAnalysis/mdanalysis/wiki/UnitTests)? (Basically, copy & paste from here together with an example usage.) This will should then also give @richardjgowers an idea how to use it in the Travis-CI test suite.

File: testsuite/MDAnalysisTests/__init__.py [hunk 1]
```diff
@@ -104,13 +104,13 @@
...
 """
 
 __version__ = "0.11.0-dev"  # keep in sync with RELEASE in setup.py
 
 try:
-    from numpy.testing import Tester, assert_, assert_equal, TestCase
-    test = Tester().test
+    from numpy.testing import assert_
 except ImportError:
     raise ImportError("""numpy>=1.5  is required to run the test suite. Please install it first. """
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/__init__.py]: Please make sure to update the numpy version requirement in both `setup.py` (package and testsuite) if we don't require 1.5+ already.

**Comment by mnmelo** [comment]
Yup. I'm waiting to see if everything builds cleanly (just saw it didn't... I'm on it!).

Regarding travis, I think we should start testing with our own script instead of the dumb `nosetests`. That will already give quieter tests and detailed knownfailure accounting.
Memleak testing is optional and off by default. Might be interesting to turn it on at some point in travis but it's probably not that relevant.

File: testsuite/MDAnalysisTests/plugins/__init__.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+# and contributors (see AUTHORS for the full list)
+#
+# Released under the GNU Public Licence, v2 or any higher version
+#
+# Please cite your use of MDAnalysis in published work:
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/plugins/__init__.py]: Add an extra comment with you as the author and year and perhaps a link to the issue.
I would like to make it extra clear that you did the whole framework.

File: testsuite/MDAnalysisTests/plugins/__init__.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+"""
+===========================
+Nose plugins for MDAnalysis
+===========================
+
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/plugins/__init__.py]: Add
```rest

:Author: Manuel Nuno Melos
:Year: 2015

.. versionadded:: 0.11.0
```

File: testsuite/MDAnalysisTests/plugins/__init__.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+Writing plugins
+===============
+
+Nose plugins can be written to expose several hooks along the nose execution
+chain. This allows very fine control of testing.
+
+The `nose plugin guidelines`_, and examples therein, are a good starting point
+on how to accomplish this. You can also base your plugins in the ones here or
+in the `nose` suite. Finally, the API specification can be found `here`_.
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/plugins/__init__.py]: I [don't like "here" links](http://www.w3.org/QA/Tips/noClickHere), instead I suggest
```rest
The `nose plugin guidelines`_, and examples therein, are a good starting point on how to accomplish this, with more details available in the `API specification`_. You can also base your plugins in the ones here or
in the `nose` suite.

.. _API specification:  http://nose.readthedocs.org/en/latest/plugins/interface.html
```

File: testsuite/MDAnalysisTests/__init__.py [hunk 5]
```diff
@@ -87,14 +87,14 @@
 The unittests use the :mod:`unittest` module together with nose_. See the
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/__init__.py]: Add an entry to `testsuite/CHANGELOG` – this is a noteworthy change in the tests!!

**Comment by orbeckst** [comment]
> Regarding travis, I think we should start testing with our own script instead of the dumb nosetests. That will already give quieter tests and detailed knownfailure accounting.
Memleak testing is optional and off by default. Might be interesting to turn it on at some point in travis but it's probably not that relevant.

I agree. Once this is merged, raise an issue for @richardjgowers to use `mda_nosetest` in travis.

Unless we want to try out `--processes=2` on travis, I think we should also test for memleaks (assuming that all tests pass...), given that we now have it. In serial, there's no real downside to memleak testing, right?

**Comment by mnmelo** [comment]
Memleak testing works fine in serial and parallel. I saw no slowndown associated with it, so I say we try it in travis and if it works leave it on, then.

I'll start working on the Wiki.

**Comment by richardjgowers** [comment]
So changing how travis runs tests is pretty simple, there's just a line  in `.travis.yml` called `script:`, I think we can just drop in `mda_nosetests` if I've understood this all right.

And there is a cryptic reference to travis having 1.5 cores per job, so maybe `--processes=2` is worth playing with too.

**Comment by mnmelo** [comment]
I will have the wiki updated soon, but yes, the idea is to use `mda_nosetests` just like `nosetests` is being used, with the addition of `--with-memleak`. I don't understad the CI environment enough to know how we can get the path to `mda_nosetests`, though.

Also, @seb-buch, will you take care of updating the version release date in the MDAnalysisTests `CHANGELOG` when 0.11 comes out?

**Pull request merged**